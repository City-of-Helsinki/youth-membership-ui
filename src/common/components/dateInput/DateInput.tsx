import React from 'react';

import DateInputGroup from './DateInputGroup';
import DateInputLogic, {
  InputComponentProps,
  DateError,
} from './DateInputLogic';
import DateInputRow from './DateInputRow';
import InputWithoutSpinners from './InputWithoutSpinners';
import styles from './dateInput.module.css';

const DEFAULT_DATE_INPUT_DATE_ID = 'ym-date-input-date-id';
const DEFAULT_DATE_INPUT_MONTH_ID = 'ym-date-input-month-id';
const DEFAULT_DATE_INPUT_YEAR_ID = 'ym-date-input-year-id';
const DEFAULT_DATE_INPUT_DATE_NAME = 'ym-date-input-date-name';
const DEFAULT_DATE_INPUT_MONTH_NAME = 'ym-date-input-month-name';
const DEFAULT_DATE_INPUT_YEAR_NAME = 'ym-date-input-year-name';
const DEFAULT_DAY_OF_MONTH_INPUT = ({
  isInvalid,
  label,
  ...rest
}: InputComponentProps) => (
  <>
    <InputWithoutSpinners
      {...rest}
      invalid={isInvalid}
      hideLabel
      labelText={label}
    />
    <span className={styles.dot}>.</span>
  </>
);
const DEFAULT_MONTH_INPUT = ({
  innerRef,
  isInvalid,
  label,
  ...rest
}: InputComponentProps) => (
  <>
    <InputWithoutSpinners
      {...rest}
      invalid={isInvalid}
      ref={innerRef}
      hideLabel
      labelText={label}
    />
    <span className={styles.dot}>.</span>
  </>
);
const DEFAULT_YEAR_INPUT = ({
  innerRef,
  isInvalid,
  label,
  ...rest
}: InputComponentProps) => (
  <InputWithoutSpinners
    {...rest}
    invalid={isInvalid}
    ref={innerRef}
    hideLabel
    labelText={label}
  />
);

interface Props {
  onChange: (date: Date) => void;
  onError: (error: DateError) => void;
  onBlur?: () => void;
  value: Date | null;
  label: string;
  error?: string;
  dateInputId?: string;
  monthInputId?: string;
  yearInputId?: string;
  dateInputName?: string;
  monthInputName?: string;
  yearInputName?: string;
  dayOfMonthComponent?: React.FC<InputComponentProps>;
  monthComponent?: React.FC<InputComponentProps>;
  yearComponent?: React.FC<InputComponentProps>;
  dateInputLabel: string;
  monthInputLabel: string;
  yearInputLabel: string;
  dateInputAutoComplete?: string;
  monthInputAutoComplete?: string;
  yearInputAutoComplete?: string;
}

function DateInput({
  onChange,
  onError,
  onBlur,
  value,
  label,
  error,
  dateInputId = DEFAULT_DATE_INPUT_DATE_ID,
  monthInputId = DEFAULT_DATE_INPUT_MONTH_ID,
  yearInputId = DEFAULT_DATE_INPUT_YEAR_ID,
  dateInputName = DEFAULT_DATE_INPUT_DATE_NAME,
  monthInputName = DEFAULT_DATE_INPUT_MONTH_NAME,
  yearInputName = DEFAULT_DATE_INPUT_YEAR_NAME,
  dayOfMonthComponent = DEFAULT_DAY_OF_MONTH_INPUT,
  monthComponent = DEFAULT_MONTH_INPUT,
  yearComponent = DEFAULT_YEAR_INPUT,
  dateInputLabel,
  monthInputLabel,
  yearInputLabel,
  dateInputAutoComplete,
  monthInputAutoComplete,
  yearInputAutoComplete,
}: Props) {
  return (
    <DateInputGroup label={label} error={error}>
      <DateInputLogic
        onChange={onChange}
        onError={onError}
        onBlur={onBlur}
        value={value}
        isInvalid={Boolean(error)}
        wrapper={DateInputRow}
        dateInputId={dateInputId}
        monthInputId={monthInputId}
        yearInputId={yearInputId}
        dateInputName={dateInputName}
        monthInputName={monthInputName}
        yearInputName={yearInputName}
        dayOfMonthComponent={dayOfMonthComponent}
        monthComponent={monthComponent}
        yearComponent={yearComponent}
        dateInputLabel={dateInputLabel}
        monthInputLabel={monthInputLabel}
        yearInputLabel={yearInputLabel}
        dateInputAutoComplete={dateInputAutoComplete}
        monthInputAutoComplete={monthInputAutoComplete}
        yearInputAutoComplete={yearInputAutoComplete}
      />
    </DateInputGroup>
  );
}

export default DateInput;
