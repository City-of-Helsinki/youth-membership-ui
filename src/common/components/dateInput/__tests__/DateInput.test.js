import React from 'react';
import toJson from 'enzyme-to-json';

import {
  render,
  // eslint-disable-next-line import/named
  fireEvent,
  // eslint-disable-next-line import/named
  getByLabelText,
} from '../../../test/testing-library';
import DateInput from '../DateInput';

describe('<DateInput />', () => {
  const dateInputLabel = 'date-input';
  const monthInputLabel = 'month-input';
  const yearInputLabel = 'year-input';
  const defaultProps = {
    value: '',
    onChange: jest.fn(),
    onError: jest.fn(),
    dateInputLabel,
    monthInputLabel,
    yearInputLabel,
  };
  const getWrapper = props =>
    render(<DateInput {...defaultProps} {...props} />);
  const dateInputSelector = container =>
    getByLabelText(container, dateInputLabel);
  const monthInputSelector = container =>
    getByLabelText(container, monthInputLabel);
  const yearInputSelector = container =>
    getByLabelText(container, yearInputLabel);

  it('should render expectedly', () => {
    expect(toJson(getWrapper())).toMatchSnapshot();
  });

  it('should yield expected dates', () => {
    const onChange = jest.fn();
    const { container } = getWrapper({ onChange });

    fireEvent.change(dateInputSelector(container), {
      target: { value: '2' },
    });
    fireEvent.change(monthInputSelector(container), {
      target: { value: '2' },
    });
    fireEvent.change(yearInputSelector(container), {
      target: { value: '2020' },
    });

    expect(onChange).toHaveBeenCalledWith(new Date(2020, 1, 2, 0, 0, 0, 0));
  });

  it('should not allow text to be inputted', () => {
    const { container, queryByDisplayValue } = getWrapper();

    const testNoTextAllowed = inputSelector => {
      const testString = 'text';

      fireEvent.change(inputSelector(container), {
        target: { value: testString },
      });

      expect(queryByDisplayValue(testString)).toEqual(null);
    };

    testNoTextAllowed(dateInputSelector);
    testNoTextAllowed(monthInputSelector);
    testNoTextAllowed(yearInputSelector);
  });

  it('should not allow negative numbers', () => {
    const { container, queryByDisplayValue } = getWrapper();

    const testNoNegativeAllowed = (inputSelector, testValue) => {
      fireEvent.change(inputSelector(container), {
        target: { value: `-${testValue}` },
      });

      expect(queryByDisplayValue(`-${testValue}`)).toEqual(null);
      expect(queryByDisplayValue(testValue.toString())).toBeDefined();
    };

    testNoNegativeAllowed(dateInputSelector, 3);
    testNoNegativeAllowed(monthInputSelector, 4);
    testNoNegativeAllowed(yearInputSelector, 2015);
  });

  it('should have a maximum character count of 2 for date and month fields, and 4 for year field', () => {
    const { container, queryByDisplayValue } = getWrapper();

    const testNoTooLongNumberAllowed = (inputSelector, numberLength) => {
      const forSureTooLongTestValue = Array.from({ length: 1000 })
        .map(() => 'a')
        .join('');

      fireEvent.change(inputSelector(container), {
        target: { value: forSureTooLongTestValue },
      });

      const expectedValue = forSureTooLongTestValue.slice(0, numberLength);

      expect(queryByDisplayValue(forSureTooLongTestValue)).toEqual(null);
      expect(queryByDisplayValue(expectedValue)).toBeDefined();
    };

    testNoTooLongNumberAllowed(dateInputSelector, 2);
    testNoTooLongNumberAllowed(monthInputSelector, 2);
    testNoTooLongNumberAllowed(yearInputSelector, 4);
  });

  it('should yield a valid date when month is january', () => {
    const onChange = jest.fn();
    const { container } = getWrapper({ onChange });

    fireEvent.change(dateInputSelector(container), {
      target: { value: '1' },
    });
    fireEvent.change(monthInputSelector(container), {
      target: { value: '1' },
    });
    fireEvent.change(yearInputSelector(container), {
      target: { value: '2017' },
    });

    expect(onChange).toHaveBeenCalledWith(new Date(2017, 0, 1, 0, 0, 0, 0));
  });

  it('should not accept invalid dates', () => {
    const onChange = jest.fn();
    const onError = jest.fn();
    const { container } = getWrapper({ onChange, onError });

    fireEvent.change(dateInputSelector(container), {
      target: { value: '31' },
    });
    fireEvent.change(monthInputSelector(container), {
      target: { value: '2' },
    });
    fireEvent.change(yearInputSelector(container), {
      target: { value: '2020' },
    });

    // Does not send invalid date
    expect(onChange.mock.calls.length).toEqual(0);
    // Sends an error instead
    expect(onError.mock.calls[0][0]).toMatchInlineSnapshot(`
      Object {
        "description": "The combination of day, month and year do not result in a date that actually exists.",
        "name": "dateDoesNotExist",
      }
    `);
  });

  it('should allow the user to input leading zeroes to day and month inputs', () => {
    const testContent = '01';
    const { container } = getWrapper();

    const testLeadingZeroesAllowed = inputSelector => {
      fireEvent.change(inputSelector(container), {
        target: { value: testContent },
      });

      expect(inputSelector(container).value).toEqual(testContent);
    };

    testLeadingZeroesAllowed(dateInputSelector);
    testLeadingZeroesAllowed(monthInputSelector);
  });
});
