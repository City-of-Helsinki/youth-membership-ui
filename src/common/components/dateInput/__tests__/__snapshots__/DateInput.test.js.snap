// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput /> should render expectedly 1`] = `
<div>
  <div
    aria-describedby="ym-date-picker-group-error"
    aria-labelledby="ym-date-picker-group"
    class="hds-text-input"
    role="group"
  >
    <span
      class="hds-text-input__label"
      id="ym-date-picker-group"
    />
    <div
      class="row"
    >
      <div
        class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq input noSpinner"
      >
        <label
          class="FieldLabel-module_label__1zrXK FieldLabel-module_hidden__3UWxI"
          for="ym-date-input-date-id"
        >
          date-input
        </label>
        <div
          class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
        >
          <input
            class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
            id="ym-date-input-date-id"
            inputmode="numeric"
            name="ym-date-input-date-name"
            pattern="\\\\d*"
            type="text"
            value=""
          />
        </div>
      </div>
      <span
        class="dot"
      >
        .
      </span>
      <div
        class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq input noSpinner"
      >
        <label
          class="FieldLabel-module_label__1zrXK FieldLabel-module_hidden__3UWxI"
          for="ym-date-input-month-id"
        >
          month-input
        </label>
        <div
          class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
        >
          <input
            class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
            id="ym-date-input-month-id"
            inputmode="numeric"
            name="ym-date-input-month-name"
            pattern="\\\\d*"
            type="text"
            value=""
          />
        </div>
      </div>
      <span
        class="dot"
      >
        .
      </span>
      <div
        class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq input noSpinner"
      >
        <label
          class="FieldLabel-module_label__1zrXK FieldLabel-module_hidden__3UWxI"
          for="ym-date-input-year-id"
        >
          year-input
        </label>
        <div
          class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
        >
          <input
            class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
            id="ym-date-input-year-id"
            inputmode="numeric"
            name="ym-date-input-year-name"
            pattern="\\\\d*"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
